### PR TITLE
Zero shot text classification pipeline

### PR DESCRIPTION
**Pull request by mgoin**
Based off of https://discuss.huggingface.co/t/new-pipeline-for-zero-shot-text-classification/681

Implements zero shot text classification pipeline. Batch size is equal to the number of sequences and currently only supports dynamic label passing, with static label support to come in a future PR. This implementation allows for future implementation of zero shot text classification based on models trained on classes other than mnli.

example dynamic labels:
```python3
zero_shot_text_classifier = Pipeline.create(
    task="zero_shot_text_classification",
    model_scheme="mnli",
    model_config={"hypothesis_template": "This text is related to {}"},     
    model_path="zoo:nlp/text_classification/distilbert-none/pytorch/"
               "huggingface/mnli/pruned80_quant-none-vnni")

sequence_to_classify = "Who are you voting for in 2020?"
candidate_labels = ["Europe", "public health", "politics"]
zero_shot_text_classifier(sequences=sequence_to_classify, labels=candidate_labels)
>>> ZeroShotTextClassificationOutput(
    sequences='Who are you voting for in 2020?',
    labels=['politics', 'public health', 'Europe'],
    scores=[0.9073666334152222, 0.046810582280159, 0.04582275450229645])
```

example static labels:
```python3
zero_shot_text_classifier = Pipeline.create(
    task="zero_shot_text_classification",
    batch_size=3,
    model_scheme="mnli",
    model_config={"hypothesis_template": "This text is related to {}"},
    model_path="zoo:nlp/text_classification/distilbert-none/pytorch/"
               "huggingface/mnli/pruned80_quant-none-vnni",
    labels=["politics", "Europe", "public health"]
)

sequence_to_classify = "Who are you voting for in 2020?"
zero_shot_text_classifier(sequences=sequence_to_classify)
>>> ZeroShotTextClassificationOutput(
    sequences='Who are you voting for in 2020?',
    labels=['politics', 'public health', 'Europe'],
    scores=[0.9073666334152222, 0.046810582280159, 0.04582275450229645])
```

Evaluation results:
Dataset sst2
Config
    multi_class = True
    hypothesis_template = "The sentiment of this text is {}"

``` bash
deepsparse.transformers.eval_downstream model_path -d sst2 --zero-shot True
```

| model | accuracy (nm pipeline) | expected accuracy(hf pipeline) |
| -------| ---------| ---------|
| [bert base](https://sparsezoo.neuralmagic.com/models/nlp%2Ftext_classification%2Fbert-base%2Fpytorch%2Fhuggingface%2Fmnli%2Fbase-none) | 0.823 | 0.823 |
| [90% pruned bert](https://sparsezoo.neuralmagic.com/models/nlp%2Ftext_classification%2Fbert-base%2Fpytorch%2Fhuggingface%2Fmnli%2F12layer_pruned90-none) | 0.779 | 0.779 |

### CODE REVIEW

**Comment by davidberenstein1957** [comment]
How far along is this PR? I am currently running a deepsparse server via the following Dockerfile, which doesn't allow me to pass 'labels' for the zero-shot classification.

`# 
FROM python:3.8-slim

# 
WORKDIR /code

# 
COPY ./requirements.txt /code/requirements.txt

# 
RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt

# 
CMD ["deepsparse.server", "--port", "80", "--task", "zero_shot_classification", "--model_path", "zoo:nlp/text_classification/distilbert-none/pytorch/huggingface/mnli/pruned80_quant-none-vnni"]
`

**Comment by davidberenstein1957** [comment]
#419

**Comment by bfineran** [comment]
hi @davidberenstein1957 this PR is still a work in progress but we do plan on merging this in the near term.  For your issue with the server, could you provide more information on how you are making the request and what error you are seeing?

For #419, thank you for the PR, it may be easier to directly rebase the branch to main here; I'll go ahead and do that so we can resolve that PR.

Happy to take any feedback you have on this pipeline and use case as well!

**Comment by davidberenstein1957** [comment]
@bfineran I want to start working with a zero-shot-classifcation model. You have conveniently trained a mnli model, which can be used for that, but as of now, the pipeline isn't supported yet. I always like pragmatic approaches, so it would be great if I could just create a simple Docker image, via the lines of code shown above.

On another matter, how do you see support for multi- and cross-lingual models coming in the future, like potentially introducing moBERT, and xlm-oBERTA?

**Comment by davidberenstein1957** [comment]
@bfineran Could you potentially just open the branch and run black formatting on the files manually?

it currently fails due to formatting errors within these 3 files.
```
would reformat src/deepsparse/tasks.py
would reformat src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
would reformat src/deepsparse/transformers/pipelines_cli.py
```

**Comment by mgoin** [comment]
@davidberenstein1957 this PR is still WIP and not ready for use yet so it's not expected to pass. Once it is in a usable state and landed, you should be able to use this pipeline through `deepsparse.server` like in your Docker example

File: src/deepsparse/transformers/pipelines/nli_text_classification.py
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/nli_text_classification.py:79]: i think `hypothesis_template` and `multi_class` makes more sense as a runtime input rather than a config. you might want to tweak this as you feed different labels and sequences to the pipeline

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:119]: can you update this?

File: src/deepsparse/transformers/pipelines/nli_text_classification.py
Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/nli_text_classification.py:79]: Since there are theoretically ways of performing zero-shot text classification with hypothesis labels (like cosine similarity, detailed here: https://joeddav.github.io/blog/2020/05/29/ZSL.html) I think hypothesis label should stay nli specific.

Multiclass I definitely agree with!

Comment (mgoin) [reply src/deepsparse/transformers/pipelines/nli_text_classification.py]: it's definitely NLI specific agreed, could it be moved to `NliTextClassificationInput`?

Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/nli_text_classification.py]: I'll make it an optional input to handle both of the cases where the hypothesis template is either static and dynamic

File: src/deepsparse/engine.py
Comment (kylesayrs) [reply src/deepsparse/engine.py:640]: @bfineran actually made a PR removing the scheduler entirely from multimodelengine, which is probably the correct design decision. I'll remove this change from here.

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:153]: could `multi_class` be a part of `NliTextClassificationInput`? i know you said before you wanted `hypothesis_template` in the config

Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:153]: I made an upstream commit here: https://github.com/neuralmagic/deepsparse/pull/450/commits/05ba598f56f122184863ffbd0995fd55e16900b6

`hypothesis_template` and `multi_class` now exist in both the nli config and nli input. If either are provided in the input, they override the default values in the config.

**Comment by mgoin** [comment]
okay thanks @kylesayrs , all good to go for this PR. i can't accept unfortunately

**Comment by kylesayrs** [comment]
Not ready for landing, I'm going to refactor the pipeline.create code to better handle the default batch size argument

File: src/deepsparse/pipeline.py
Comment (bfineran) [inline src/deepsparse/pipeline.py:740]:
```suggestion
    transformers zero shot text classification pipeline
```

File: src/deepsparse/transformers/pipelines/nli_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/nli_text_classification.py:37]: given the purpose of these helper functions is for shared functionality across potential future pipelines, and these helper functions assume properties of the given pipelines such as `_threadpool`, would it make sense to refactor this into a parent nli pipeline class that handles the processing as well as creation of the threadpool, config, etc?

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (mgoin) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:209]: can you leave a TODO or asana task for this `2`? We will want to revisit this value when the elastic scheduler has a form of work-stealing across streams

File: src/deepsparse/transformers/pipelines/nli_text_classification.py
Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/nli_text_classification.py:37]: I think the vision was to allow for users to choose between nli and future implementations of zero shot via the model_scheme constructor arg. I'm not aware of any ways to dynamically inherit from a parent class at construction time.

If anything it seems like the nli class should inherit from a zero shot pipeline class, but I believe that implementation will break once new child classes are added, since they can't all be registered under the same "zero_shot_text_classification" task.

After investigating a bit more, perhaps we can override the `__new__` method, which will allow the construction of the `ZeroShotTextClassificationPipeline` to dynamically decide to return a `MnliTextClassificationPipeline`. The mnli pipeline will then inherit from the zero shot pipeline and override its own methods.

Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/nli_text_classification.py:37]:
```
class ZeroShotTextClassificationPipeline():
    def __new__(cls, model_scheme):
        print(f"ZeroShotTextClassificationPipeline __new__")

        if model_scheme == "mnli":
            return MnliTextClassificationPipeline()
        elif model_scheme == "mlm":
            return MlmTextClassificationPipeline()
        else:
            raise ValueError()

class MlmTextClassificationPipeline(TransformersPipeline):
    # cannot inherit from ZeroShotTextClassificationPipeline
    pass

class MnliTextClassificationPipeline(TransformersPipeline):
    # cannot inherit from ZeroShotTextClassificationPipeline
    pass
```

**Comment by kylesayrs** [comment]
Just finished refactoring (https://github.com/neuralmagic/deepsparse/pull/369#discussion_r912411628). Ready for review

File: src/deepsparse/transformers/pipelines/mnli_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/mnli_text_classification.py:233]: nit: use double quotes

**Review by bfineran — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/mnli_text_classification.py
Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/mnli_text_classification.py:133]: This was requested by https://github.com/neuralmagic/deepsparse/pull/369#discussion_r911481686

Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/mnli_text_classification.py:133]: Is there a better way to designate places where future implementation will be needed?

Comment (kylesayrs) [reply src/deepsparse/transformers/pipelines/mnli_text_classification.py]: Removed and folded into dynamic batch changes

**Comment by kylesayrs** [comment]
TODO: Performance testing

File: src/deepsparse/utils/data.py
Comment (bfineran) [inline src/deepsparse/utils/data.py:162]: max and sum are bad names (python built-ints)

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:211]:
```suggestion
class ZeroShotTextClassificationPipelineBase(TransformersPipeline):
```

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:181]:
```suggestion
class ZeroShotTextClassificationInputBase(BaseModel):
```

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:249]: if the config class is already defined at this point, typing should point to the config class not `BaseModel`, if it is not defined yet, should be the config class name in quotes. `BaseModel` typing only used when accepting generic schema objects

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:247]: let's move parse_config and parse_labels to bottom of this class' functions to allow for base `Pipeline` overrides to come first

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:35]: add documentation on intended relationship between zero-shot pipeline, bouncer class, and implementations

File: src/deepsparse/transformers/pipelines/mnli_text_classification.py
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/mnli_text_classification.py:135]: note: this will need to be refactored after #491 lands

**Review by bfineran — CHANGES_REQUESTED** [review]
looks good pending open comments,  let's add an option to run this pipeline in `eval_downstream` (can be follow up PR)

**Review by bfineran — DISMISSED** [review]

**Review by bfineran — DISMISSED** [review]
great work @kylesayrs approving now that we have eval confirmation

File: src/deepsparse/transformers/eval_downstream.py
Comment (corey-nm) [inline src/deepsparse/transformers/eval_downstream.py:50]:
```suggestion
                        Currently supports zero shot pipelines for the `sst2` dataset.
```

File: src/deepsparse/transformers/pipelines/zero_shot_text_classification.py
Comment (corey-nm) [inline src/deepsparse/transformers/pipelines/zero_shot_text_classification.py:264]:
```suggestion
```

😁

**Review by corey-nm — DISMISSED** [review]

**Review by bfineran — APPROVED** [review]

**Review by corey-nm — APPROVED** [review]

**Comment by kylesayrs** [comment]
@davidberenstein1957 This has been merged and is undergoing QA testing

**Pull request merged**
